scripts/29To35Rewritten.R

```diff
@@ -0,0 +1,142 @@
+# Assignment 1:  
+library(tweedie) 
+library(ggplot2)
+#running the doParallel library
+library(doParallel)
+
+#Starting a parallel backend with the number of cores available
+cl <- makeCluster(detectCores() - 1) # leave one core free
+registerDoParallel(cl)
+
+simTweedieTest <-  
+  function(N){ 
+    t.test( 
+      rtweedie(N, mu=10000, phi=100, power=1.9), 
+      mu=10000 
+    )$p.value 
+  } 
+
+
+# Assignment 2:  
+MTweedieTests <-  
+  function(N,M,sig){ 
+    sum(replicate(M,simTweedieTest(N)) < sig)/M 
+  } 
+
+
+# Assignment 3:  
+df <-  
+  expand.grid( 
+    N = c(10,100,1000,5000, 10000), 
+    M = 1000, 
+    share_reject = NA) 
+
+#the for loop rewritten to a for each loop
+foreach(
+  i = 1:nrow(df), 
+  df$share_reject[i] <-  
+    MTweedieTests( 
+      N=df$N[i], 
+      M=df$M[i], 
+      sig=.05) 
+)
```

Comment:
This one is not completely correct. When using the foreach() function, you have to use th %dopar% to initialize the paralleziation. Look at the alternative solution below:

```r
# Next we use doParallel for the loop over the result df: 
maxcores <- 8
Cores <- min(parallel::detectCores(), maxcores)
cl <- makeCluster(Cores)
registerDoParallel(cl)


res <-
  foreach(
    i = 1:nrow(df),
    .combine = 'rbind',
    .packages = c('magrittr', 'dplyr', 'tweedie')
  ) %dopar%
  tibble(
    N = df$N[i],
    M = df$M[i],
    share_reject =
      MTweedieTests(N = df$N[i],
                    M = df$M[i],
                    sig = 0.05)
  )
```